File: curryfire/UIView+TwelvePrinciples.m
//
//  UIView+TwelvePrinciples.m
//  Created by Devin Ross on 4/22/15.
//
/*
 
 curryfire || https://github.com/devinross/curry-fire
 
 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:
 
 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 */

#import "UIView+TwelvePrinciples.h"
#import "ShortHand.h"


@implementation UIView (TwelvePrinciples)


- (void) zoomToYPoint:(CGFloat)endYPoint completion:(void (^)(BOOL finished))completion{
    [self zoomToYPoint:endYPoint duration:0.7 delay:0 completion:completion];
}


- (void) zoomToYPoint:(CGFloat)endYPoint duration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
	[self zoomToYPoint:endYPoint anticipation:30 duration:duration delay:delay completion:completion];
}

- (void) zoomToYPoint:(CGFloat)endYPoint anticipation:(CGFloat)anticipation duration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
	
	CGAffineTransform baseTransform = self.transform;
	
	if(endYPoint > self.centerY){
		CGFloat yy = CGRectGetMinY(self.frame);
		self.layer.anchorPoint = CGPointMake(0.5, 0);
		self.center = CGPointMake(self.centerX, yy);
		anticipation *= -1;
	}else{
		CGFloat yy = CGRectGetMaxY(self.frame);
		self.layer.anchorPoint = CGPointMake(0.5, 1);
		self.center = CGPointMake(self.centerX, yy);
	}
	
	
	CGFloat xScale = 1, yScale = 1;
	
	[UIView animateKeyframesWithDuration:duration delay:delay options:UIViewKeyframeAnimationOptionCalculationModeCubic animations:^{
		
		[UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.2 animations:^{
			self.transform = CGConcat(baseTransform, CGScale(xScale+0.1, yScale-0.3));
		}];
		
	
		
		[UIView addKeyframeWithRelativeStartTime:0.3 relativeDuration:0.3 animations:^{
			self.transform = CGScale(xScale-0.3, yScale+0.3);
		}];
		
		[UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.4 animations:^{
			self.centerY += anticipation;
		}];
		
		[UIView addKeyframeWithRelativeStartTime:0.4 relativeDuration:0.6 animations:^{
			self.center = CGPointMake(self.centerX, endYPoint);
		}];
		
	} completion:^(BOOL finished){
		self.transform = baseTransform;
		CGFloat minY = self.minY + self.height / 2.0f;
		self.layer.anchorPoint = CGPointMake(0.5, 0.5);
		self.centerY = minY;
		if(completion) completion(finished);
	}];
}



- (void) zoomToXPoint:(CGFloat)endXPoint completion:(void (^)(BOOL finished))completion{
    [self zoomToXPoint:endXPoint completion:completion];
}
- (void) zoomToXPoint:(CGFloat)endXPoint duration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
    
    CGAffineTransform baseTransform = self.transform;

    if(endXPoint > self.centerX){
        CGFloat xx = CGRectGetMinX(self.frame);
        self.layer.anchorPoint = CGPointMake(0, 0.5);
        self.center = CGPointMake(xx, self.centerY);
    }else{
        CGFloat xx = CGRectGetMaxX(self.frame);
        self.layer.anchorPoint = CGPointMake(1, 1);
        self.center = CGPointMake(xx, self.centerY);
    }
    
    CGFloat xScale = self.transform.a;
    CGFloat yScale = self.transform.d;
    
    [UIView animateKeyframesWithDuration:duration delay:delay options:UIViewKeyframeAnimationOptionCalculationModeCubic animations:^{
        
        [UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.2 animations:^{
            self.transform = CGScale(xScale-0.3, yScale+0.1);
        }];
        
        [UIView addKeyframeWithRelativeStartTime:0.3 relativeDuration:0.3 animations:^{
            self.transform = CGScale(xScale+0.3, yScale-0.3);
        }];
        
        [UIView addKeyframeWithRelativeStartTime:0.4 relativeDuration:0.6 animations:^{
            self.center = CGPointMake(endXPoint, self.centerY);
        }];
        
    } completion:^(BOOL finished){
        self.transform = baseTransform;

        CGFloat minX = self.minX + self.width / 2.0f;
        self.layer.anchorPoint = CGPointMake(0.5, 0.5);
        self.centerX = minX;
        if(completion) completion(finished);
    }];
}


- (void) hop{
	[self hopWithToXPoint:self.superview.width * 1.5 hopHeight:40 duration:0.7 delay:0 completion:nil];
}
- (void) hopWithToXPoint:(CGFloat)xPoint hopHeight:(CGFloat)hopHeight duration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
	
	CGFloat y = self.center.y;
	CGAffineTransform baseTransform = self.transform;
	
	CGPoint midPoint = CGPointGetMidpoint(self.center, CGPointMake(xPoint, self.centerY));
	
	[UIView animateKeyframesWithDuration:duration delay:delay options:UIViewKeyframeAnimationOptionCalculationModeCubic animations:^{
		
		[UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.3 animations:^{
			self.transform = CGConcat(CGScale(0.8, 1), CGRotate(-15 * M_PI / 180));
		}];
		[UIView addKeyframeWithRelativeStartTime:0.3 relativeDuration:0.3 animations:^{
			self.center = CGPointMake(midPoint.x, y - hopHeight);
			self.transform = CGConcat(CGScale(1.4, 1), CGRotate(0 * M_PI / 180));
		}];
		[UIView addKeyframeWithRelativeStartTime:0.6 relativeDuration:0.4 animations:^{
			self.center = CGPointMake(xPoint, y);
		}];
		[UIView addKeyframeWithRelativeStartTime:0.6 relativeDuration:0.2 animations:^{
			self.transform = CGConcat(CGScale(1.1, 1), CGRotate(15 * M_PI / 180));
		}];
		[UIView addKeyframeWithRelativeStartTime:0.8 relativeDuration:0.2 animations:^{
			self.transform = baseTransform;
		}];
		
	}completion:completion];
	
}



- (void) turnOnADimeAtXPoint:(CGFloat)endXPoint duration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
	
	BOOL movingRight = endXPoint > self.centerX;

	CGFloat anchorX = movingRight ? self.maxX : self.minX;
	CGFloat anchorY = self.maxY;
	CGFloat mult = movingRight ? 1 : -1;
	self.layer.anchorPoint = movingRight ? CGPointMake(1, 1) : CGPointMake(0, 1);
	
	CGFloat twist = movingRight ? 0.1 : -0.1;
	CGAffineTransform twistTransform = CGAffineTransformIdentity;
	twistTransform.c = twist;
	
	CGFloat x = endXPoint + (movingRight ? 1 : -1) * self.width / 2;
	self.center = CGPointMake(anchorX, anchorY);
	
	[UIView animateKeyframesWithDuration:duration delay:delay options:UIViewKeyframeAnimationOptionCalculationModeCubic animations:^{
		
		double s = 0.0, dur = 0.3;
		
		[UIView addKeyframeWithRelativeStartTime:s relativeDuration:dur/2 animations:^{
			self.transform = twistTransform;
		}];
		
		[UIView addKeyframeWithRelativeStartTime:s + dur/2 relativeDuration:dur/2 animations:^{
			CGAffineTransform transform = CGAffineTransformIdentity;
			transform.c = 0;
			self.transform = transform;
		}];
		
		
		[UIView addKeyframeWithRelativeStartTime:s relativeDuration:dur animations:^{
			self.center = CGPointMake(x + 20 * mult, anchorY);
		}];
		
		s += dur;
		dur = 0.2;
		
		[UIView addKeyframeWithRelativeStartTime:s relativeDuration:dur animations:^{
			CGAffineTransform transform = CGAffineTransformIdentity;
			transform.c = 0;
			transform = CGConcat(transform, CGRotate(-2 * M_PI / 180.0f));
			self.transform = transform;
			self.center = CGPointMake(x + 30 * mult, anchorY-16);
		}];
		
		s += dur;
		dur = 0.2;
		
		[UIView addKeyframeWithRelativeStartTime:s relativeDuration:dur animations:^{
			CGAffineTransform transform = CGAffineTransformIdentity;
			transform.c = 0;
			transform = CGConcat(transform, CGRotate(-1 * M_PI / 180.0f));
			self.transform = transform;
			self.center = CGPointMake(x + 15 * mult, anchorY-20);
		}];
		
		s += dur;
		dur = 0.2;
		
		[UIView addKeyframeWithRelativeStartTime:s relativeDuration:dur  animations:^{
			CGAffineTransform transform = CGAffineTransformIdentity;
			transform = CGConcat(transform, CGRotate(0 * M_PI / 180.0f));
			self.transform = transform;
			self.center = CGPointMake(x, anchorY);
		}];
		
		
	} completion:^(BOOL finished) {
		
		
		CGFloat xx = CGRectGetMidX(self.frame), yy = CGRectGetMidY(self.frame);
		
		self.layer.anchorPoint = CGPointMake(0.5, 0.5);
		self.center = CGPointMake(xx,yy);
		
		if(completion)
			completion(finished);
		
	}];

	
}




- (void) shakeAnimationWithDuration:(NSTimeInterval)duration delay:(NSTimeInterval)delay completion:(void (^)(BOOL finished))completion{
    
    CGFloat centerY = self.centerY, centerX = self.centerX;
    
    [UIView animateKeyframesWithDuration:0.6 delay:0 options:UIViewKeyframeAnimationOptionCalculationModeCubicPaced animations:^{
        
        [UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.2 animations:^{
            self.center = CGPointMake(centerX - 15, centerY);
        }];
        [UIView addKeyframeWithRelativeStartTime:0.2 relativeDuration:0.2 animations:^{
            self.center = CGPointMake(centerX + 15, centerY);
        }];
        [UIView addKeyframeWithRelativeStartTime:0.4 relativeDuration:0.2 animations:^{
            self.center = CGPointMake(centerX - 10, centerY);
        }];
        [UIView addKeyframeWithRelativeStartTime:0.6 relativeDuration:0.2 animations:^{
            self.center = CGPointMake(centerX + 10, centerY);
        }];
        [UIView addKeyframeWithRelativeStartTime:0.6 relativeDuration:0.2 animations:^{
            self.center = CGPointMake(centerX + 0, centerY);
        }];
        
    } completion:completion];

}
- (void) shakeAnimationWithCompletion:(void (^)(BOOL finished))completion{
    [self shakeAnimationWithDuration:0.6 delay:0 completion:completion];
}


- (void) wiggle{
    [self wiggleWithRotation:0.06];
}
- (void) wiggleWithRotation:(CGFloat)angle{
    [self wiggleWithDuration:0.5 delay:0 rotation:angle];
}
- (void) wiggleWithDuration:(NSTimeInterval)duration delay:(NSTimeInterval)delay betweenAngleOne:(CGFloat)angleOne angleTwo:(CGFloat)angleTwo{
    
    CGAffineTransform transform = self.transform;
    CGFloat currentAngle = atan2(transform.b, transform.a);
    CABasicAnimation *animation;
    
    CATransform3D second = CATransform3DMakeRotation(angleTwo, 0, 0, 1.0);
    CFTimeInterval secondBeginTime;

    if(angleOne == currentAngle || angleTwo == currentAngle){
        if(angleTwo == currentAngle)
            second = CATransform3DMakeRotation(angleOne, 0, 0, 1.0);
        secondBeginTime = CACurrentMediaTime() + delay;
    }else{
        
        CATransform3D first = CATransform3DMakeRotation(angleOne, 0, 0, 1.0);
        animation = [CABasicAnimation animationWithKeyPath:@"transform"];
        animation.fromValue = NSCATransform3D(self.layer.transform);
        animation.toValue = NSCATransform3D(first);
        animation.duration = duration / 2.0;
        animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
        animation.removedOnCompletion = NO;
        animation.beginTime = CACurrentMediaTime() + delay;
        animation.fillMode = kCAFillModeForwards;
        [self addAnimation:animation];
        secondBeginTime = CACurrentMediaTime() + duration / 2.0 + delay;
    }
    
    animation = [CABasicAnimation animationWithKeyPath:@"transform"]; 	// Create a basic animation to animate the layer's transform
    animation.toValue = NSCATransform3D(second); 	// Assign the transform as the animation's value
    animation.autoreverses = YES;
    animation.duration = duration;
    animation.repeatCount = HUGE_VALF;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    animation.beginTime = secondBeginTime;
    animation.removedOnCompletion = NO;
    animation.fillMode = kCAFillModeForwards;
    [self addAnimation:animation];
    
    
}
- (void) wiggleWithDuration:(NSTimeInterval)duration delay:(NSTimeInterval)delay rotation:(CGFloat)angle{
    [self wiggleWithDuration:duration delay:delay betweenAngleOne:angle angleTwo:-angle];
}

- (void) runForrestRunToPoint:(CGPoint)point withCompletion:(void (^)(BOOL finished))completion{
    [self runForrestRunWithDuration:1 delay:0 toPoint:point completion:completion];
}
- (void) runForrestRunWithDuration:(NSTimeInterval)duration delay:(NSTimeInterval)delay toPoint:(CGPoint)point completion:(void (^)(BOOL finished))completion{
	
	
	BOOL movingRight = point.x > self.centerX;
	
    CGAffineTransform baseTransform = self.transform;
	CGAffineTransform transform = baseTransform;
	transform.c = movingRight ? 0.5 : -0.5;

	CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform"];
	animation.autoreverses = YES;
	animation.fromValue = [NSValue valueWithCATransform3D:CATransform3DMakeAffineTransform(baseTransform)];
	animation.toValue = [NSValue valueWithCATransform3D:CATransform3DMakeAffineTransform(transform)];
	animation.duration = duration / 4;
	[self.layer addAnimation:animation forKey:@"forrest"];
    
	
    [UIView animateWithDuration:duration delay:0 usingSpringWithDamping:0.7 initialSpringVelocity:0.4 options:0 animations:^{
        self.center = point;
    }completion:completion];
    
}

@end

File: curryfire/UIView+TwelvePrinciples.swift
//
//  UIView+TwelvePrinciples.swift
//  Created by Devin Ross on 9/13/16.
//
/*

curryfire || https://github.com/devinross/curry-fire

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

*/

import UIKit


extension UIView {
	
	/// Hop the view.
	@objc public func hop() {
		self.hop(self.superview!.width * 1.5, hopHeight: 40, duration: 0.7, delay: 0, completion: { _ in })
	}
	
	/** Hop the view.
	@param xPoint The place to hop.
	@param hopHeight The height of the hop.
	@param duration The duration of the animation.
	@param delay The delay before the animation is played.
	@param completion The completion callback handler.
	*/
	@objc public func hop(_ xPoint: CGFloat, hopHeight: CGFloat, duration: TimeInterval, delay: TimeInterval, completion: ((Bool) -> Void)?) {
		let y: CGFloat = self.center.y
		let baseTransform = self.transform
		let midPoint = CGPointGetMidpoint(self.center, CGPoint(x: xPoint,y: self.centerY))
		UIView.animateKeyframes(withDuration: duration, delay: delay, options: .calculationModeCubic, animations: {() -> Void in
			UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 0.3, animations: {() -> Void in
				self.transform = CGConcat(CGScale(0.8, 1.0), CGRotate((-15.0 * CGFloat.pi) / 180.0))
			})
			UIView.addKeyframe(withRelativeStartTime: 0.3, relativeDuration: 0.3, animations: {() -> Void in
				self.center = CGPoint(x: midPoint.x, y: y - hopHeight)
				self.transform = CGConcat(CGScale(1.4, 1), CGRotate((0.0 * CGFloat.pi) / 180.0))
			})
			UIView.addKeyframe(withRelativeStartTime: 0.6, relativeDuration: 0.4, animations: {() -> Void in
				self.center = CGPoint(x: xPoint, y: y)
			})
			UIView.addKeyframe(withRelativeStartTime: 0.6, relativeDuration: 0.2, animations: {() -> Void in
				self.transform = CGConcat(CGScale(1.1, 1), CGRotate(15.0 * CGFloat.pi / 180.0))
			})
			UIView.addKeyframe(withRelativeStartTime: 0.8, relativeDuration: 0.2, animations: {() -> Void in
				self.transform = baseTransform
			})
			}, completion: completion)
	}
	
	
	
	
	/** Shake the view.
	 @param withDuration The duration of the animation.
	 @param delay The delay before the animation is played.
	 @param completion The completion callback handler.
	 */
	@objc public func shake(_ withDuration: TimeInterval, delay: TimeInterval, completion: ((Bool) -> Void)? ) {
		let centerY: CGFloat = self.centerY
		let centerX: CGFloat = self.centerX
		UIView.animateKeyframes(withDuration: 0.6, delay: 0, options: .calculationModeCubicPaced, animations: {() -> Void in
			UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 0.2, animations: {() -> Void in
				self.center = CGPoint(x: centerX - 15, y: centerY)
			})
			UIView.addKeyframe(withRelativeStartTime: 0.2, relativeDuration: 0.2, animations: {() -> Void in
				self.center = CGPoint(x: centerX + 15, y: centerY)
			})
			UIView.addKeyframe(withRelativeStartTime: 0.4, relativeDuration: 0.2, animations: {() -> Void in
				self.center = CGPoint(x: centerX - 10, y: centerY)
			})
			UIView.addKeyframe(withRelativeStartTime: 0.6, relativeDuration: 0.2, animations: {() -> Void in
				self.center = CGPoint(x: centerX + 10, y: centerY)
			})
			UIView.addKeyframe(withRelativeStartTime: 0.6, relativeDuration: 0.2, animations: {() -> Void in
				self.center = CGPoint(x: centerX + 0, y: centerY)
			})
			}, completion: completion)
	}
	
	/** Shake the view.
	@param completion The completion callback handler.
	*/
	@objc public  func shake(_ completion: ((Bool) -> Void)? ) {
		self.shake(0.6, delay: 0, completion: completion)
	}
	
	
	
	
	
	
	/** Run forrest. Run.
	 @param toPoint The place run to.
	 @param withCompletion The completion callback handler.
	 */
	@objc public func runForrestRun(_ toPoint: CGPoint, withCompletion completion: ((Bool) -> Void)?) {
		self.runForrestRun(1, delay: 0, toPoint: toPoint, completion: completion)
	}
	
	
	/** Run forrest. Run.
	 @param duration The duration of the animation.
	 @param delay The delay before the animation is played.
	 @param point The place run to.
	 @param completion The completion callback handler.
	 */
	@objc public func runForrestRun(_ duration: TimeInterval, delay: TimeInterval, toPoint point: CGPoint, completion: ((Bool) -> Void)?) {
		let movingRight = point.x > self.centerX
		let baseTransform = self.transform
		var transform = baseTransform
		transform.c = movingRight ? 0.5 : -0.5
		let animation = CABasicAnimation(keyPath: "transform")
		animation.autoreverses = true
		animation.fromValue = NSValue(caTransform3D: CATransform3DMakeAffineTransform(baseTransform))
		animation.toValue = NSValue(caTransform3D: CATransform3DMakeAffineTransform(transform))
		animation.duration = duration / 4
		self.layer.add(animation, forKey: "forrest")
		UIView.animate(withDuration: duration, delay: 0, usingSpringWithDamping: 0.7, initialSpringVelocity: 0.4, options: [], animations: {() -> Void in
			self.center = point
			}, completion: completion)
	}

	
	
	
	/** Stop on a dime
	@param endXPoint The duration of the animation.
	@param duration The duration of the animation.
	@param delay The delay before the animation is played.
	@param completion The completion callback handler.
	*/
	@objc public func turnOnADime(_ endXPoint: CGFloat, duration: TimeInterval, delay: TimeInterval, completion: ((Bool) -> Void)?) {
		let movingRight = endXPoint > self.centerX
		let anchorX: CGFloat = movingRight ? self.maxX : self.minX
		let anchorY: CGFloat = self.maxY
		let mult: CGFloat = movingRight ? 1 : -1
		self.layer.anchorPoint = movingRight ? CGPoint(x: 1, y: 1) : CGPoint(x: 0, y: 1)
		let twist: CGFloat = movingRight ? 0.1 : -0.1
		var twistTransform = CGAffineTransform.identity
		twistTransform.c = twist
		let x: CGFloat = endXPoint + (movingRight ? 1 : -1) * self.width / 2
		self.center = CGPoint(x: anchorX, y: anchorY)
		
		

		
		UIView.animateKeyframes(withDuration: duration, delay: delay, options: .calculationModeCubic, animations: { () -> Void in
			var s: Double = 0.0
			var dur: Double = 0.3
			UIView.addKeyframe(withRelativeStartTime: s, relativeDuration: dur / 2, animations: {() -> Void in
				self.transform = twistTransform
			})
			UIView.addKeyframe(withRelativeStartTime: s + dur / 2, relativeDuration: dur / 2, animations: {() -> Void in
				var transform = CGAffineTransform.identity
				transform.c = 0
				self.transform = transform
			})
			UIView.addKeyframe(withRelativeStartTime: s, relativeDuration: dur, animations: {() -> Void in
				self.center = CGPoint(x: x + 20 * mult, y: anchorY)
			})
			s += dur
			dur = 0.2
			UIView.addKeyframe(withRelativeStartTime: s, relativeDuration: dur, animations: {() -> Void in
				var transform = CGAffineTransform.identity
				transform.c = 0
				transform = CGConcat(transform, CGRotate(-2 * CGFloat.pi / 180.0))
				self.transform = transform
				self.center = CGPoint(x: x + 30 * mult, y: anchorY - 16)
			})
			s += dur
			dur = 0.2
			UIView.addKeyframe(withRelativeStartTime: s, relativeDuration: dur, animations: {() -> Void in
				var transform = CGAffineTransform.identity
				transform.c = 0
				transform = CGConcat(transform, CGRotate(-1 * CGFloat.pi / 180.0))
				self.transform = transform
				self.center = CGPoint(x: x + 15 * mult, y: anchorY - 20)
			})
			s += dur
			dur = 0.2
			
			
			
			UIView.addKeyframe(withRelativeStartTime: s, relativeDuration: dur, animations: {() -> Void in
				var transform = CGAffineTransform.identity
				transform = CGConcat(transform, CGRotate(0 * CGFloat.pi / 180.0))
				self.transform = transform
				self.center = CGPoint(x: x, y: anchorY)
			})
			
		}, completion: { (finished) in
			let xx: CGFloat = self.frame.midX
			let yy: CGFloat = self.frame.midY
			self.layer.anchorPoint = CGPoint(x: 0.5, y: 0.5)
			self.center = CGPoint(x: xx, y: yy)
			if (completion != nil) {
				completion!(finished)
			}
		})
	}
}
